teps/0007-conditions-beta.md

```diff
@@ -0,0 +1,610 @@
+
+### Simplicity
+
+As discussed in the background section, `Conditions` manifest themselves as `Tasks`. We want to keep Tekton as simple as possible by reusing existing components. So we propose phasing out the separate `Conditions` CRD and eventually deprecating it. In place of `Conditions`, we want to use `Tasks` produce to `Results` that we can use to specify `Guards` using `When Expression` in subsequent `Tasks`, as described in [Efficiency](#efficiency) section below. Thus, we won’t have `Conditions` to migrate to beta and won’t have to maintain the separate `Conditions` CRD. 
+
+In the example of checking whether a file exists, the `Task` that would replace the `Condition` would be specified as such:
+
+```yaml
+apiVersion: tekton.dev/v1alpha1
+kind: Task
+metadata:
+  name: file-exists
+spec:
+  params:
+    - name: path
+  workspaces:
+    - name: source
+  results:
+    - name: exists
+      description: boolean indicating whether the file exists
+  steps:
+  - name: check-file-exists
+    image: alpine
+    script: |
+      if [ -f $(workspaces.source.path)/$(params.path) ]; then
+        echo true | tee /tekton/results/exists
+      else
+        echo false | tee /tekton/results/exists
+      fi
+```
+
+### Efficiency
+To improve the efficiency of guarded execution of `Tasks`, we need to avoid spinning up new `Pods` to check `Guards`. We can use string expressions for `Guard` checking, but we want to avoid adding an opinionated and complex expression language to the Tekton API to ensure Tekton can be supported by as many systems as possible. We propose using a simple expression syntax `When Expressions` (similar to Kubernetes' `When Expressions`) to evaluate `Guards` efficiently. The components of `When Expressions` are `Input`, `Operator` and `Values`. `Input` is the input for the `Guard` checking which can be static inputs or outputs from parent `Tasks`, such as `Parameters` or `Results`. `Values` is an array of string values. `Operator` represents an `Input`'s relationship to a set of `Values`. `Operators` we will use in `Guards` are `In` and `NotIn`. The `Values` array must be non-empty. When we have more than one `Guard`, the guarded `Task` will be executed when all the `Guards` evaluate to `True`. Note that when a `Guard` uses a resource from another `Task`, such as a `Result`, it introduces an implicit resource dependency that makes the guarded `Task` dependent on the resource-producing `Task`. 
```

Comment:
could you include a link to the kubernetes docs and/or code on their match expressions?

(also nitpick but could you add some newlines in this paragraph when the lines get long - makes it easier to comment on specific parts of the paragraph at least :D). it also might be easier to read as several paragraphs, e.g. current looks like:

![image](https://user-images.githubusercontent.com/432502/89219201-591ee480-d59d-11ea-9287-a4a721f00de6.png)

might be easier to read as something like:

The components of When Expressions are Input, Operator and Values:
 * Input is the input for the Guard checking which can be static inputs or outputs from parent Tasks, such as Parameters or Results. 
 * ...